apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -685,24 +705,42 @@ async function traverseQuestionAndExecuteFunctions(phase, codeCaller, data, cont
         }
       }
     }
+
+    if (!('childNodes' in node)) {
+      // This is a text node or comment node; there are no children to process.
+      return node;
+    }
+
+    /** @type {parse5.DefaultTreeAdapterTypes.ChildNode[]} */
     const newChildren = [];
-    for (let i = 0; i < (node.childNodes || []).length; i++) {
-      const childRes = await visitNode(node.childNodes[i]);
-      if (childRes) {
-        if (childRes.nodeName === '#document-fragment') {
-          newChildren.push(...childRes.childNodes);
-        } else {
-          newChildren.push(childRes);
-        }
+    for (const childNode of node.childNodes) {
+      const childRes = await visitNode(childNode);
+
+      // We can't rely on `nodeName` as a discriminant because for the `Element`
+      // type, it doesn't have a literal value:
+      // https://github.com/microsoft/TypeScript/issues/48500#issuecomment-1085063454
+      // We use custom type guards instead.
+
+      if (isDocumentFragmentNode(childRes)) {
+        newChildren.push(...childRes.childNodes);
+      } else if (!isDocumentNode(childRes)) {
```

Comment:
This is a new condition that wasn't there before. Would this cause issues if a question contains something that is interpreted as a document (whatever that means)? I assume no, but I just want to make sure this is considered.

Reply:
I was actually able to eliminate the need for this check entirely in 06914d5. How does that look?